src/Npgsql/NpgsqlConnector.FrontendMessages.cs

```diff
@@ -0,0 +1,420 @@
+
+            foreach (var kv in parameters)
+            {
+                WriteBuffer.WriteString(kv.Key);
+                WriteBuffer.WriteByte(0);
+                WriteBuffer.WriteString(kv.Value);
+                WriteBuffer.WriteByte(0);
+            }
+
+            WriteBuffer.WriteByte(0);
+        }
+
+        #region Authentication
+
+        internal Task WritePassword(byte[] payload, bool async) => WritePassword(payload, 0, payload.Length, async);
+
+        internal async Task WritePassword(byte[] payload, int offset, int count, bool async)
+        {
+            if (WriteBuffer.WriteSpaceLeft < 1 + 5)
+                await WriteBuffer.Flush(async);
+            WriteBuffer.WriteByte((byte)'p');
+            WriteBuffer.WriteInt32(4 + count);
+
+            if (count <= WriteBuffer.WriteSpaceLeft)
+            {
+                // The entire array fits in our WriteBuffer, copy it into the WriteBuffer as usual.
+                WriteBuffer.WriteBytes(payload, offset, count);
+                return;
+            }
+
+            await WriteBuffer.Flush(async);
+            WriteBuffer.DirectWrite(payload, offset, count);
+        }
+
+        internal async Task WriteSASLInitialResponse(string mechanism, byte[] initialResponse, bool async)
+        {
+            var len = 1 + 4 +
+                      PGUtil.UTF8Encoding.GetByteCount(mechanism) + 1 +
+                      4 + (initialResponse?.Length ?? 0);
+
+            if (WriteBuffer.WriteSpaceLeft < len)
+                await WriteBuffer.Flush(async);
+
+            WriteBuffer.WriteByte((byte)'p');
+            WriteBuffer.WriteInt32(len - 1);
+
+            WriteBuffer.WriteString(mechanism);
+            WriteBuffer.WriteByte(0);   // null terminator
+            if (initialResponse == null)
+                WriteBuffer.WriteInt32(-1);
+            else
+            {
+                WriteBuffer.WriteInt32(initialResponse.Length);
+                WriteBuffer.WriteBytes(initialResponse);
+            }
+        }
+
+        internal Task WriteSASLResponse(byte[] payload, bool async) => WritePassword(payload, async);
+
+        #endregion Authentication
+
+        internal Task WritePregenerated(byte[] data, bool async=false)
+        {
+            if (WriteBuffer.WriteSpaceLeft < data.Length)
+                return FlushAndWrite(data, async);
+
+            WriteBuffer.WriteBytes(data, 0, data.Length);
+            return Task.CompletedTask;
+
+            async Task FlushAndWrite(byte[] data, bool async)
+            {
+                await Flush(async);
+                Debug.Assert(data.Length <= WriteBuffer.WriteSpaceLeft, $"Pregenerated message has length {data.Length} which is bigger than the buffer ({WriteBuffer.WriteSpaceLeft})");
+                WriteBuffer.WriteBytes(data, 0, data.Length);
+            }
+        }
+
+        internal void Flush() => WriteBuffer.Flush(false).GetAwaiter().GetResult();
```

Comment:
Is this ever called?

Reply:
Yep, from the COPY code, as well as several other places where we don't have an async path. Do you see an issue?

Reply:
Ok, no issue then. Was reviewing on my cellphone and must have overlooked the usages.